Consistently structure vendor exports + use compatibility helper

nikku opened the pull request:
### Proposed Changes

This PR ensures we follow a consistent structure for vendor exports, and per file core component exports:

```
import someComponent from 'camunda-modeler-plugin-helpers/vendor/...';
import Fill from 'camunda-modeler-plugin-helpers/components/Fill.js';
```

#### React exports now exposed via `vendor`

The previous `camunda-modeler-plugin-helpers/react.js` must be adjusted:

```diff
- import React from 'camunda-modeler-plugin-helpers/react.js';
+ import React from 'camunda-modeler-plugin-helpers/vendor/react.js';
```

#### Core components now exposed via individual exports

Core components are no longer exposed via `camunda-modeler-plugin-helpers/components.js`, but as dedicated (per component) exports:

```diff
- import { Fill, Modal } from 'camunda-modeler-plugin-helpers/components.js';
+ import Fill from 'camunda-modeler-plugin-helpers/components/Fill.js';
+ import Modal from 'camunda-modeler-plugin-helpers/components/Modal.js';
```

#### Incompatibility is indicated early

It also ensures we consistently throw, eagerly, with human readable error, if the host modeler does not provide a given dependency we rely on. For that purpose we re-use the existing `returnOrThrow(fn, minimumCompatVersion)` utility.


### Checklist

To ensure you provided everything we need to look at your PR:

* [x] __Brief textual description__ of the changes present
* [ ] __Visual demo__ attached
* [ ] __Steps to try out__ present, i.e. [using the `@bpmn-io/sr` tool](https://github.com/bpmn-io/sr)
* [ ] Related issue linked via `Closes {LINK_TO_ISSUE}` or `Related to {LINK_TO_ISSUE}`

<!--
Thanks for creating this pull request! ❤️
-->

nikku committed:
chore: gracefully handle missing `vendor` exports

1. adjusted properties-panel related functionality in vendor
2. expose `react` under `vendor`
    * added deprecated react warning
    * added test case that verify vendor/react works
3. chore: expose `@carbon/*` under `vendor`
    * added test case that verify vendor/carbon works

nikku committed:
chore!: expose individual components and expose components folder in package.json

This allows us to version (and handle incompatibility)
of various components individually. It also improves
how the plug-in components are being loaded, and how
errors are handled (immediately!).

Previously:

* Using a non-existing component would throw an error
  late, once the plug-in instantiates

Now:

* Accessing a non-existing component throws immediately

BREAKING CHANGE: Use components exported via `camunda-modeler-plugin-helpers/components/*`.

barmac commented:
I think this makes sense. We have structured the library to use `/vendor` once, and I don't see a reason why Carbon would be an exception.

philippfromme commented:
This will also require a major release of https://github.com/camunda/camunda-modeler-webpack-plugin/, right?

So the scenario for me as a plugin maintainer as I understand it is (option **CURRENT**):

* no changes in upcoming Desktop Modeler (e.g., `window.react` unchanged)
* if I don't update to new major version of `camunda-modeler-plugin-helpers` I don't have to do anything
* if I update to new major version of `camunda-modeler-plugin-helpers`
  * if I'm using React through `window.React` directly, good for me 🤡 
  * if I'm using React through `import React from "camunda-modeler-plugin-helpers/react"` I will get a module not found error
  * if I'm using React through `import React from "react"` without updating the Webpack plugin or my Webpack config as well I will also get a module not found error
  * in both cases there is no deprectation warning

The alternative for us would be:

  A. Don't move helpers
  B. Move helpers but add deprectation warning

```javascript
// camunda-modeler-plugin-helpers/react.js

console.log('Deprecated and will be removed future versions');

export { default } from './vendors/react';
```

barmac commented:
Good catch with React @philippfromme! Indeed this PR is a major. Since the webpack plugin requires plugin helpers 5.x, there is no need to major-release the webpack plugin. It only needs to release a version which supports the new major of the plugin helpers.

abdul99ahad commented:
I agree with @philippfromme, adding deprecation warning seems like a way to go.

nikku commented:
Thanks for your comments. I'd be fine to keep the existing exports, but add a compatibility warning. I think this will pave the path towards a better future, and "support our users" during the transition. 

Note that the way components were structured in the past, some issues would not be found during plug-in loading, but [during rendering](https://github.com/camunda/camunda-modeler-plugin-helpers/pull/31/files#diff-9952085201d5294bbe13e9af4011a32c3534cf1cff58505c1a0ec59ac4a3757cL9). This is something that is different now (incompatibility is indicated early). We cannot ship that part as a minor, unfortunately.

So what we could do (instead), option **C**: 

* Blow up when using `components.js` or `react.js`, with a human readable error
* Release as major

Let's decide, and then I can update the PR accordingly.

----

Unrelated to this PR, as you found @philippfromme, there is a strong connection between the webpack plug-in and this library. Probably the plug-in should declare a peer dependency to this library, so incompatibility is caught early.

nikku commented:
Updated the PR to be more descriptive.

nikku commented:
https://github.com/camunda/camunda-modeler-plugin-helpers/pull/31/commits/13fe3dfec7e11665e00449be9fc5a6bbc884fd6e implements option **C**, which is to indicate misuse with an error:

```
camunda-modeler-plugin-helpers/components.js is no longer provided. Migrate to individual component exports: https://github.com/camunda/camunda-modeler-plugin-helpers/pull/31
```

```
camunda-modeler-plugin-helpers/react.js is no longer provided. Migrate to use the respective camunda-modeler-plugin-helpers/vendor/react.js export: https://github.com/camunda/camunda-modeler-plugin-helpers/pull/31
```

Alternative: 

* Fail during bundling (rather than at run-time), which would be option **CURRENT**

abdul99ahad commented on vendor/react.js:
```diff
@@ -1,3 +1,8 @@
+ *
+ * @type {import('react')}
+ */
+export default returnOrThrow(() => window.react, '3.4');
```
If we are exporting via vendor shoudn't it be exposed by window.vendor?.react (ref. https://github.com/camunda/camunda-modeler/pull/4860#issuecomment-2711429915)

```suggestion
export default returnOrThrow(() => window.vendor?.react, '3.4');
```

so in modeler, when it calls for react package from helper, it will look into vendor object to see if it has react package setup.

nikku replied:
No, we cannot do this, as it would be a major (Camunda Modeler) platform breaking change.

Of course, we could start to export `react` from `window.vendor` moving forward, but this will break _any_ existing plug-in.

barmac commented:
> Probably the plug-in should declare a peer dependency to this library, so incompatibility is caught early.

It does! https://github.com/camunda/camunda-modeler-webpack-plugin/blob/main/package.json#L61

barmac commented on components.js:
```diff
@@ -1,257 +1 @@
- */
-export const createTab = window.components.createTab || NotCompatible('5.29');
\ No newline at end of file
+throw new Error('camunda-modeler-plugin-helpers/components.js is no longer provided. Migrate to individual component exports: https://github.com/camunda/camunda-modeler-plugin-helpers/pull/31');
```
Shift left! I like it.

nikku replied:
But shift left would be:

* Remove the exports, so this fails during build time already.

nikku replied:
As per "further shift left" I removed https://github.com/camunda/camunda-modeler-plugin-helpers/commit/13fe3dfec7e11665e00449be9fc5a6bbc884fd6e again.

philippfromme replied:
Good solution!

nikku replied:
[fyi] Removed the "good solution" again, cf. https://github.com/camunda/camunda-modeler-plugin-helpers/pull/31#discussion_r1989280609.

We want plug-in authors to fail as early as possible ("shift left", as @barmac frames it).

philippfromme commented:
Is this a mistake?

![image](https://github.com/user-attachments/assets/5fd5c620-3870-4641-b9ff-a083f731fdce)

All of our components are exported from `components`, right?

barmac commented:
Before we merge anything, I'd like us to try this out on an existing UI plugin, e.g. [the autosave plugin](https://github.com/pinussilvestrus/camunda-modeler-autosave-plugin)

nikku commented:
Updated [changelog](https://github.com/camunda/camunda-modeler-plugin-helpers/pull/31/commits/cf6686b0a250c8216e73cf7860fb7198c2ebc2cb).

nikku commented:
> https://github.com/camunda/camunda-modeler-plugin-helpers/pull/31#issuecomment-2714759967

@philippfromme You mean in the PR description? Yes, a documentation bug, fixed in PR description.

nikku commented on react.js:
```diff
@@ -0,0 +1,3 @@
+console.log('Warning: This module is deprecated and will be removed in future versions. Please use the module from ./vendors/react instead.');
```
```suggestion
console.log('Warning: This module is deprecated and will be removed in future versions. Please use the module exposed via "camunda-modeler-plugin-helpers/vendors/react.js" instead.');
```

nikku replied:
Let's use this with extension, and use the full export.

nikku replied:
Also, consider to log this as an error, so a clear stack trace is provided.

nikku replied:
Given that this change is major breaking anyway I don't see a strong reason to include this additional, informative hint.

We want to shift left, and trying to use `camunda-modeler-plugin-helpers/react.js` will yield a bundling error. Cf. https://github.com/camunda/camunda-modeler-plugin-helpers/pull/31#discussion_r1989506551.

abdul99ahad replied:
resolved.

abdul99ahad commented on vendor/react.js:
```diff
@@ -1,3 +1,8 @@
+ *
+ * @type {import('react')}
+ */
+export default returnOrThrow(() => window.react, '3.4');
```
We can get default exports with this but named exports are throwing error when tested on `auto-save-plugin`

There are 2 things we can do.
1. Switch back to cjs 
2. We can use something like this 

```js
import React from 'camunda-modeler-plugin-helpers/vendor/react';
const { useState , useEffect } = React;
```
or maybe 3rd, a even better way?

nikku replied:
Named exports throw an error when importing?

abdul99ahad replied:
when importing `import React, { useState, useEffect } from 'camunda-modeler-plugin-helpers/vendor/react'`, the named imports are not found.

 `export 'useState' (imported as 'useState') was not found in 'camunda-modeler-plugin-helpers/vendor/react' (possible exports: default)`

nikku replied:
Do we have a test case to verify this behavior? I guess we should have one.

nikku replied:
I'll have a look.

abdul99ahad replied:
> Do we have a test case to verify this behavior? I guess we should have one.

We have in webpack plugin added via https://github.com/camunda/camunda-modeler-webpack-plugin/pull/8

nikku replied:
I looked into this and, indeed, this there is no simple solution to this but to destructure the window object and explicitly re-export it. I'll see if I can create a test case to verify this scenario (as part of this library), next week.

abdul99ahad replied:
I can take this over if you want. 

What I'm thinking of is the test case which verifies the react object from window. (we can mock window, maybe). There can be two cases, 
1. where it's trying to import without destructuring, throws error. 
```js
import React, { useState } from 'camunda-modeler-plugin-helpers/vendor/react' //throws error
```
2. The other one destructure the imports from package object (in this case, react)
```js
import React from 'camunda-modeler-plugin-helpers/vendor/react'
const { useState } = React;
```

WDYT?

nikku replied:
I think you're right on the test cases. I think we should verify that both things work, whatever direction we go.

https://github.com/camunda/camunda-modeler-plugin-helpers/pull/31/commits/e76ea450f683677cb449144a91e22e7f56985ae0 should be green, when tested against our normal setup.

abdul99ahad commented:
update: The test case is not working as expected. It requires react object before importing. I'm currently working to resolve this so we can close this PR. 

![image](https://github.com/user-attachments/assets/c77ab891-1d40-4ee5-bf6f-463496367b82)

abdul99ahad commented:
I've fixed the test cases default and named exports. You can find it here 2cdeb969434d93a715b4b61583ed111a2ba66d91. Awaiting feedback.

abdul99ahad commented:
This PR looks good to me now. I'll test it with the plugin to ensure everything works as expected. If no further changes are needed, feel free to mark it as ready for review and I will merge.

This will unblock https://github.com/camunda/camunda-modeler-plugin-helpers/pull/29, and we can add carbon imports and release major version :) 

cc: @nikku

nikku commented on test/spec/vendor/react.spec.js:
```diff
@@ -1,13 +1,39 @@
 describe('vendor/react', function() {
 
-  it('should export', function() {
+  let React, useState, useEffect;
```
What about the following:

* We have a dedicated helper to mock `window` exports, set it up before all test runs; this way we can ensure that the imports actually work.
* Alternatively, we make this a "bundling only" test, as long as bundling works (without warnings) we're fine
* Alternatively we resort to your way of testing, but I'm not sure we can get along with "only default export should be used". This is out for a long time, users may already consume it via named imports.

```
import React, { useState } from 'camunda-modeler-plugin-helpers/vendor/react.js';
```

nikku replied:
The [plug-in](https://github.com/camunda/camunda-modeler-webpack-plugin), for example, only replaces `react` with the plug-in helper import. A named react import will break the application.

abdul99ahad replied:
> I'm not sure we can get along with "only default export should be used". This is out for a long time, users may already consume it 

That's why we will release this as BREAKING CHANGE. we discussed this [here](https://github.com/camunda/camunda-modeler-plugin-helpers/pull/31#discussion_r2003700846)

> The [plug-in](https://github.com/camunda/camunda-modeler-webpack-plugin), for example, only replaces react with the plug-in helper import. A named react import will break the application.

We're expecting to only give complete react module via `plugin-helpers` and users can get further exports like (useState, useEffect,...) from exported object. 
https://github.com/camunda/camunda-modeler-plugin-helpers/blob/2cdeb969434d93a715b4b61583ed111a2ba66d91/test/spec/vendor/react.spec.js#L29-L32

this is what this test case verifies.

abdul99ahad replied:
> Alternatively, we make this a "bundling only" test, as long as bundling works (without warnings) we're fine

We can add these tests in `web-pack-plugin` as integration tests as this helper doesn't bundle.

nikku replied:
Hmm, but why should our export not export exactly what `React` exports? I don't understand. Our export should be a drop-in replacement, no?

So what is 

```
import React, { useState } from 'react';
```

would become

```
import React, { useState } from 'camunda-modeler-plugin-helpers/vendor/react.js';
```

What keeps us from accomplishing that?

abdul99ahad replied:
> What keeps us from accomplishing that?

I didn't dig deep enough in this issue but the [usage of ES6](https://github.com/camunda/camunda-modeler-plugin-helpers/pull/28) within the lib seems to be creating these issues. 
![image](https://github.com/user-attachments/assets/e7e2f4cd-6558-406b-b701-eea3e46ee45c)

Whereas, when working with earlier version (cjs), this was working fine. Maybe, this has to do something with default exports in es6. 

When I import named imports from the default 
```
import { React } from 'camunda-modeler-plugin-helpers/vendor/react';
const { useState } = React;
```
This was working fine.

nikku replied:
So then, let's tackle it either way:

* We explicitly export everything that shall be exported from the upstream library (mirror their ES6 exports)
* _Or_ we keep these files as CJS, knowing that it would work then

abdul99ahad commented:
I'm taking over this PR as part of carbon integration epic.

abdul99ahad commented:
**Update**: I have reverted back to cjs. I have tried this on [the autosave plugin](https://github.com/pinussilvestrus/camunda-modeler-autosave-plugin) and it's working fine. 

### Steps to try
1. checkout this branch and run `npm ci`
2. clone [autosave plugin](https://github.com/pinussilvestrus/camunda-modeler-autosave-plugin). You might need to adjust component imports.
3. Link `camunda-modeler-plugin-helpers#compatible` to `autosave plugin`
4. Run `npm run all` on the plugin. It should bundle fine.

### Next steps 

I will check with carbon plugin and see how carbon imports behave.

abdul99ahad committed:
chore: revert back to cjs

reason: https://github.com/camunda/camunda-modeler-plugin-helpers/pull/31#discussion_r2030290351, https://stackoverflow.com/questions/40294870/module-exports-vs-export-default-in-node-js-and-es6

cjs exports entire react package and esm exports defaults and named exports. we needed complete react and carbon package.

abdul99ahad commented:
Reverted back to cjs. Components and configurations are still ESM. Didn't find the reason why it should also migrate to cjs. This PR is ready for review. Steps to try out can be found in the comment before. Awaiting feedback

abdul99ahad committed:
chore(CHANGELOG): update to v6.0.0

nikku commented on vendor/@carbon/icons-react.js:
```diff
@@ -0,0 +1,8 @@
+ *
+ * @type {import('@carbon/icons-react')}
+ */
+module.exports = returnOrThrow(() => window.carbonicons, '5.33');
```
Should we change to `window.vendor.carbonicons` to align with other exports, i.e. `preact` and `properties-panel`?

abdul99ahad replied:
Adjusted all other imports via `vendor` object https://github.com/camunda/camunda-modeler-plugin-helpers/pull/31/commits/91a2544e03b3908c80f1e41b20817e295f1f7265

nikku replied:
@abdul99ahad the implication of that change is that we need to fix upstream (camunda-modeler) and adjust the RPA plug-in, right? I just want to ensure you own the big picture. Just because I ask a question does not mean that we need to go the asked for route :).

abdul99ahad replied:
No, the RPA plugin doesn't use `plugin-helpers` so we won't need to change `camunda-modeler`. If plugin author is using the `plugin-helpers` and `webpack` correctly, we just need to adjust aliasing and global references of packages i.e., React, Carbon from global object to vendor object.

nikku replied:
And RPA also does not use any of our existing exports? `window.carbon*`?

abdul99ahad replied:
Nope, it just uses the styles.

nikku commented:
This looks good to me, thanks :star:

jarekdanielak commented:
I tried it with my [forked autosave plugin](https://github.com/jarekdanielak/camunda-modeler-autosave-plugin) and it throws:

> Error: Not compatible with Camunda Modeler < 3.4

The imports are:

```js
import React, { useEffect, useState, useRef } from 'camunda-modeler-plugin-helpers/vendor/react';
import Fill from 'camunda-modeler-plugin-helpers/components/Fill';
```

Am I missing something?

abdul99ahad commented:
> I tried it with my [forked autosave plugin](https://github.com/jarekdanielak/camunda-modeler-autosave-plugin) and it throws:
Error: Not compatible with Camunda Modeler < 3.4

We discussed this and since these have **BREAKING CHANGES**, we have to adjust the modeler.

jarekdanielak commented on vendor/@carbon/react.js:
```diff
@@ -0,0 +1,8 @@
+ *
+ * @type {import('@carbon/react')}
+ */
+module.exports = returnOrThrow(() => window.vendor?.carbonReact, '5.33');
```
```suggestion
module.exports = returnOrThrow(() => window.vendor?.carbonReact, '5.38');
```

It is only going to be available in the future version.

jarekdanielak commented on vendor/@carbon/icons-react.js:
```diff
@@ -0,0 +1,8 @@
+ *
+ * @type {import('@carbon/icons-react')}
+ */
+module.exports = returnOrThrow(() => window.vendor?.carbonIconsReact, '5.33');
```
```suggestion
module.exports = returnOrThrow(() => window.vendor?.carbonIconsReact, '5.38');
```

jarekdanielak commented:
Let's clean up the commits before merging.

jarekdanielak reviewed (approved):
Looks good. Let's clean up the commits & merge.
